### ISSUE TITLE
OSPRay 1.1.1: the Nth directional lights, for N > 1 do not appear to work correctly

### ISSUE THREAD

**Issue by tachyon-john** [ospray/OSPRay#93]
When I specify multiple directional lights, it seems that OSPRay is ignoring any light except for the first directional light.  AFAIK I'm adding them correctly, it's pretty simple code, and I can see that the values being set are all good going in, so I'm puzzled as to why the second and subsequent lights seem to be ignored?

  //
  // Set lights
  //
  for (int i = 0; i < directional_lights.num(); ++i) {
    OSPLight light = ospNewLight(ospRenderer, "DirectionalLight");

    // AO scaling factor is applied at the renderer level, but
    // we apply the direct lighting scaling factor to the lights.
    ospSet3f(light, "color",
             directional_lights[i].color[0] * ao_direct,
             directional_lights[i].color[1] * ao_direct,
             directional_lights[i].color[2] * ao_direct);

    float lightDir[3];
    vec_copy(lightDir, directional_lights[i].dir);
    vec_normalize(lightDir); // OSPRay requires normalized light direction

    // OSPRay uses a light direction vector opposite to VMD and Tachyon
    ospSet3f(light, "direction", -lightDir[0], -lightDir[1], -lightDir[2]);

    ospCommit(light);
    ospLights.push_back(light);
  }
  ospLightData = ospNewData(ospLights.size(), OSP_OBJECT, &ospLights[0], OSP_DATA_SHARED_BUFFER);

[... then at a later point ...]

ospSetData(ospRenderer, "lights", ospLightData);

**Comment by johguenther** [comment]
I cannot reproduce the problem. However, with the `OSP_DATA_SHARED_BUFFER` flag for `ospLightData` I get segfaults – so please try without sharing the lights buffer.

Some more hints:
- Instead of multiplying `ao_direct` to the light color you can also just set the light "`intensity`" parameter to `ao_direct`.
- Normalizing the light direction is not (no longer?) necessary, this is always done internally by the directional light.

**Comment by tachyon-john** [comment]
After replacing the OSP_DATA_SHARED_BUFFER with 0, I still do not observe correct behavior.  It is as if the other lights are simply not getting added, or that their contributions are nulled out.  It isn't a direction issue because I have rotated them all the way around 360 degrees checking at various positions to ensure that something weird wasn't going on.  

The suggestion to use the new "intensity" parameter works fine for the one light that does work, so I'm putting that change in... :-)

**Comment by johguenther** [comment]
Are all the lights bright enough? There is currently a hardcoded contribution threshold of 0.01.

To check that all lights are indeed known to the renderer, can you `PRINT(lightArray.size())` at the end of `ospray/render/scivis/SciVisRenderer.cpp::commit()`?

Similarly, can you please `PRINT(radiance)` at the end of `ospray/lights/DirectionalLight.cpp::commit()`?

If this does not help, is there a way to make a reproducer available?

**Comment by tachyon-john** [comment]
The lights all have intensity set to 1.0 in the simplest cases where I also observe the problem.  I don't have a build tree setup for compiling OSPRay from source at this moment, but I can dig into that when I get back from SC.

**Comment by tachyon-john** [comment]
Johannes, I have tracked this down to a combination of two issues on my side.  Since some very early version of the code, the "Ns" material property had been getting written to as a float3 rather than as a float, causing specular highlights to be rendered badly.  I also found a spot where a light direction array subscript was zero rather than following the light index.  Fixing these two issues I now see the lights behaving much closer to what I expect.  I'm now investigating why the scene still appears somewhat darker than I would expect when AO is off, but looks entirely correct  when AO is on.
